NXP IRQSTEER support for ARM arch

yangbolu1991 opened the pull request:
This PR is to add ARM arch support for NXP IRQSTEER.
Verified on i.MX95 M7.

yangbolu1991 committed:
arch: arm: cortex_m: handle multi-level interrupts in irq init

Multi-level interrupts should be handled in irq init. Only level1
interrupts going directly to interrupt controller need to be
initialized.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: interrupt_controller: intc_nxp_irqsteer: support ARM Cortex-M

Added ARM Cortex-M support for intc_nxp_irqsteer driver.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: interrupt_controller: intc_nxp_irqsteer: fix system INTID

Current system INTID calcualtion only worked for SoCs whose extended
interrupts started from IRQ 0.
Otherwise, FSL_FEATURE_IRQSTEER_IRQ_START_INDEX should be added for
system INTID.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

dleach02 commented on arch/arm/core/cortex_a_r/irq_manage.c:
```diff
@@ -40,17 +40,17 @@ extern void z_arm_reserved(void);
 
 #if !defined(CONFIG_ARM_CUSTOM_INTERRUPT_CONTROLLER)
-void arch_irq_enable(unsigned int irq)
+void arm_irq_enable(unsigned int irq)
```
These sets of changes to the function names shouldn't be done. I understand that this is under the arm folder but if you look at all the rest of the irq_manage files in the other architectures they use the generic arch_irq naming

wearyzen replied:
+1 please do not change the function names.

yangbolu1991 replied:
As we know, current arm irq.h gives interrupt control functions clarification arch_irq_*.

- For ARM_CUSTOM_INTERRUPT_CONTROLLER case, this allows custom interrupt controller implementation.
- For non-ARM_CUSTOM_INTERRUPT_CONTROLLER case, we just use arch_irq_* functions in irq_manage.c.

So, what does the case of multi-level interrupt (ARM interrupt controller + SoC extended interrupts via ARM interrupt lines) belong to?

This case can't use arch_irq_* functions in irq_manage.c directly and also not ARM_CUSTOM_INTERRUPT_CONTROLLER neither.

If arch_irq_* intend to control whole soc interrupts, I think it should cover multi-level interrupt case. And supported in irq.h as you see CONFIG_MULTI_LEVEL_INTERRUPT I added in irq.h patch.
irq control via NVIC/GIC may be exported as unified APIs arm_irq_* so that hardware module driver extending interrupts could use to implement custom z_soc_irq_*.

That's what I did in PR. If it's not the way to support, do you have any suggestion?
Thank you very much.

wearyzen replied:
For starters, lets revert the changes you made in irq_manage.c and irq.h. Since, you don't define CONFIG_ARM_CUSTOM_INTERRUPT_CONTROLLER you should be able to just define 
`#define irqsteer_arch_irq_enable(irq) irq_enable(irq);`
Keep the changes in irq_init.c for the review, I'll get back to you on that.

yangbolu1991 replied:
Thanks @wearyzen 
I updated the PR to v2 per suggestion. Changes include,
- Dropped one patch for arm irq control functions reworking.
- Converted to use irq_*() functions in irqsteer driver for ARM.

I am not still not clear how to do next. These functions here actually are expected for level 1 interrupt control.
`#define irqsteer_arch_irq_enable(irq) irq_enable(irq);`

Hope for your guide for next step.
Thank you very much.

yangbolu1991 replied:
@wearyzen 
Any suggestion for next?
Thanks a lot.

wearyzen replied:
Hi @yangbolu1991, after some digging I found [irq_next_level_api](https://github.com/zephyrproject-rtos/zephyr/blob/50c21f159143e18c8a05cbb946a5e7610033355f/include/zephyr/irq_nextlevel.h#L53).

These are the example intc drivers that use isr_nextlevel_api:
- https://github.com/zephyrproject-rtos/zephyr/blob/50c21f159143e18c8a05cbb946a5e7610033355f/drivers/interrupt_controller/intc_dw.c#L142
- https://github.com/zephyrproject-rtos/zephyr/blob/50c21f159143e18c8a05cbb946a5e7610033355f/drivers/interrupt_controller/intc_cavs.c#L113
- https://github.com/zephyrproject-rtos/zephyr/blob/50c21f159143e18c8a05cbb946a5e7610033355f/drivers/interrupt_controller/intc_rv32m1_intmux.c#L138


IIUC using `irq_next_level_api`'s is the right way to do irq initialization for interrupt aggregator and with this you won't have to change the `z_arm_interrupt_init` or any arch specific files. But I don't have much experience with these api's myself so I would let other experts to comment on that.
Meanwhile I would recommend looking at these apis and checking if they work for your case.

yangbolu1991 replied:
Thanks @wearyzen 
This is not the problem we should address first.
The first problem to address is to allow customed irq control with ARM interrupt controller (multi-level interrupt) in arm irq.h. Then the second problem is if we need to use irq_next_level_api to make the customed irq control.

Because in arm irq.h, current irq control is clarified as below.
```
#if !defined(CONFIG_ARM_CUSTOM_INTERRUPT_CONTROLLER)
<here it's forced to control level 1 interrupt of GIC and NVIC. we can't custom anything.>
#else
<here ARM interrupt controller is disabled. we can't support muli-leve interrupt.>
#endif
```

So in my v1 PR, actually my proposal was,
```
#if defined(CONFIG_ARM_CUSTOM_INTERRUPT) || defined (CONFIG_MULTI_LEVEL_INTERRUPTS)
<here we allow customed irq control even we use ARM interrupt controller. this is for multi-level interrupt. Since we still need to control ARM interrupt controller during custom, I exported arm irq control APIs (for NVIC/GIC control).>
#else
<here we still force to control level 1 interrupt of GIC and NVIC.>
#endif
```

Thanks. Hope for suggestion for the first problem.

wearyzen replied:
We already discussed the problems with previous patch, (renaming the arch_irq apis and calling arm_irq apis from driver). 

> The first problem to address is to allow customed irq control with ARM interrupt controller (multi-level interrupt) in arm irq.h

At the time I was thinking the 2nd level interrupt init would be done in [irqsteer_init](https://github.com/zephyrproject-rtos/zephyr/blob/e885c8c42f775f92176e7ad7eb71d8dcdc2bb4e0/drivers/interrupt_controller/intc_nxp_irqsteer.c#L585) but please give me some more time to look at this. Also, there is [this](https://github.com/zephyrproject-rtos/zephyr/issues/16168) which suggests that multi level interrupt handling is not supported for a while (but I may be wrong).
I'll get back to you once I have something. Meanwhile please feel free to request other reviewers for the suggestions as well.

wearyzen replied:
We already discussed the problems with previous patch, (renaming the arch_irq apis and calling arm_irq apis from driver). 

> The first problem to address is to allow customed irq control with ARM interrupt controller (multi-level interrupt) in arm irq.h

At the time I was thinking the 2nd level interrupt init would be done in [irqsteer_init](https://github.com/zephyrproject-rtos/zephyr/blob/e885c8c42f775f92176e7ad7eb71d8dcdc2bb4e0/drivers/interrupt_controller/intc_nxp_irqsteer.c#L585) but I missed your comment in z_arm_interrupt_init() and thought it was doing something more. The changes in z_arm_interrupt_init look fine to me now, I have update my comment over there.

dleach02 reviewed (dismissed):
Don't rename the arch_irq functions

wearyzen commented on include/zephyr/arch/arm/irq.h:
```diff
@@ -34,17 +34,11 @@ GTEXT(z_soc_irq_eoi)
 #else
 
-#if !defined(CONFIG_ARM_CUSTOM_INTERRUPT_CONTROLLER)
+extern void arm_irq_enable(unsigned int irq);
```
It doesn't make sense to use Arm arch_irq_*() APIs if CONFIG_ARM_CUSTOM_INTERRUPT_CONTROLLER. Why can't the `drivers/interrupt_controller/intc_nxp_irqsteer.c` define these 3 APIs?

yangbolu1991 replied:
The intc_nxp_irqsteer case now is arm interrupt controller + irqsteer module extending interrupts.
arm irq.h is proper place to export arm interrupt control functions. Then driver no longer need to care about the interrupt controller is NVIC or GIC or anyelse in the future.

Thanks.

yangbolu1991 commented on drivers/interrupt_controller/intc_nxp_irqsteer.c:
```diff
@@ -265,6 +265,18 @@ LOG_MODULE_REGISTER(nxp_irqstr);
+#define irqsteer_arch_irq_disable(irq)    xtensa_irq_disable(XTENSA_IRQ_NUMBER(irq));
+#define irqsteer_arch_irq_is_enabled(irq) xtensa_irq_is_enabled(XTENSA_IRQ_NUMBER(irq));
+#elif defined(CONFIG_ARM)
+#define irqsteer_arch_irq_enable(irq)     arm_irq_enable(irq);
```
This is to support SoC multi-level interrupts based on ARM interrupt controller.
arch_irq_* need to be customed, as previous XTENSA ARCH did.
This kind case may be not considered in arm irq.h previously.

wearyzen commented on arch/arm/core/cortex_m/irq_init.c:
```diff
@@ -27,7 +27,12 @@ void z_arm_interrupt_init(void)
 {
 	int irq = 0;
 
+/* CONFIG_2ND_LVL_ISR_TBL_OFFSET could be treated as level1 interrupt number */
```
Xtensa does this differently [here](https://github.com/zephyrproject-rtos/zephyr/blob/50c21f159143e18c8a05cbb946a5e7610033355f/include/zephyr/arch/xtensa/irq.h#L78) but I would still recommend looking at the irq_nextlevel_api instead of fixing this.

yangbolu1991 replied:
Xtensa is different. 
NUM_IRQS is Kconfig option for ARM while it is macro for xtensa in header file.
So, soc defines NUM_IRQS value in soc Kconfig.
Thanks.

wearyzen replied:
You are right, I missed the comment above the loop. So CONFIG_NUM_IRQS is total number of irqs and we are just initializing the level1 irqs here which looks fine.

wearyzen commented on arch/arm/core/cortex_m/irq_init.c:
```diff
@@ -27,7 +27,12 @@ void z_arm_interrupt_init(void)
 {
 	int irq = 0;
 
+/* CONFIG_2ND_LVL_ISR_TBL_OFFSET could be treated as level1 interrupt number */
```
```suggestion
/* CONFIG_2ND_LVL_ISR_TBL_OFFSET could be treated as total number of level1 interrupts */
```

yangbolu1991 replied:
okay, thanks.

wearyzen commented:
BTW, there is also [this](https://github.com/zephyrproject-rtos/zephyr/issues/16168) which suggests that multi level interrupt handling is not supported for a while (I may be wrong) so just wanted to know how did you test this?
Do you plan to add any new test to cover the multi level use-case?
Meanwhile please feel free to request other reviewers for the suggestions as well.

yangbolu1991 committed:
arch: arm: irq: allow custom irq control for multi-level interrupts

Custom irq control for multi-level interrupts should be allowed.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
> @yangbolu1991, thank you for the explanation. BTW, there is also [this](https://github.com/zephyrproject-rtos/zephyr/issues/16168) which suggests that multi level interrupt handling is not supported for a while (I may be wrong) so just wanted to know how did you test this? Do you plan to add any new test to cover the multi level use-case? Meanwhile please feel free to request other reviewers for the suggestions as well.

@wearyzen  I hadn't had plan to add any new test.
Actually multilevel interrupt test is already covered by the IP tests whose platform enable multilevel interrupt. Because if the interrupt is not working, the IP function is not working either.
As I mentioned this PR is a dependency for https://github.com/zephyrproject-rtos/zephyr/pull/81113,
it's working for i.MX95 platform NET.

@LaurentiuM1234  Could you help to view the PR?
Thanks

yangbolu1991 commented:
Updated to v3. Changes inlcudes,
- @wearyzen I reworked the ARM IRQ in another way to support multilevel interrupt, as arch_irq names were not expected to be changes. Just exported arm irq control functions. Also added more annotation in code for understanding.

Thanks a lot.

wearyzen commented on arch/arm/core/cortex_m/irq_manage.c:
```diff
@@ -91,6 +110,27 @@ void z_arm_irq_priority_set(unsigned int irq, unsigned int prio, uint32_t flags)
 	NVIC_SetPriority((IRQn_Type)irq, prio);
 }
 
+#if !defined(CONFIG_MULTI_LEVEL_INTERRUPTS)
```
the version 2 of the patch already looked good so I don't understand why we need these changes?

yangbolu1991 replied:
No actually :(
The v2 patch was not workable. I had thought you might give me more suggestion to modify based on v2.

wearyzen commented on drivers/interrupt_controller/intc_nxp_irqsteer.c:
```diff
@@ -265,6 +265,18 @@ LOG_MODULE_REGISTER(nxp_irqstr);
+#define irqsteer_level1_irq_disable(irq)    xtensa_irq_disable(XTENSA_IRQ_NUMBER(irq))
+#define irqsteer_level1_irq_is_enabled(irq) xtensa_irq_is_enabled(XTENSA_IRQ_NUMBER(irq))
+#elif defined(CONFIG_ARM)
+#define irqsteer_level1_irq_enable(irq)     arm_irq_enable(irq)
```
why can't this just be? :
```suggestion
#define irqsteer_level1_irq_enable(irq)     irq_enable(irq)
```

yangbolu1991 replied:
There may be some misunderstanding for you. Let me clarify.

irq_enable() should be irq control for whole soc covering multi-level interrupts which is used everywhere in drivers.
The calling is irq_enable() -> arch_irq_enable().

What irqsteer driver did is to define the custom irq_enable/arch_irq_enable with z_soc_irq_enable() to support multi-level interrupts.
So, irq_enable() can not be used. The driver needs level 1 interrupt control API to implement z_soc_irq_enable().

See current logic in arm irq.h.

```
#if !defined(CONFIG_ARM_CUSTOM_INTERRUPT_CONTROLLER)
extern void arch_irq_enable(unsigned int irq);
#else
void z_soc_irq_init(void);
#define arch_irq_enable(irq)            z_soc_irq_enable(irq)
#endif
```

What we need and the patch did is to change to the below logic.
Otherwise, irq_enable is forced to NVIC/GIC APIs which is not able to support multi-level interrupt.

```
#if !defined(CONFIG_ARM_CUSTOM_INTERRUPT_CONTROLLER)
extern void arch_irq_enable(unsigned int irq);
#endif

#if defined(CONFIG_ARM_CUSTOM_INTERRUPT_CONTROLLER) || defined(CONFIG_MULTI_LEVEL_INTERRUPTS)
void z_soc_irq_init(void);
#define arch_irq_enable(irq)            z_soc_irq_enable(irq)
#endif
```

Thanks.

yangbolu1991 replied:
@wearyzen ,
I understand you. But that's not workable.
Level 2 interrupt control is finally through controlling the corresponding level 1 interrupt. 
So, here z_soc_irq_level2_enable(irq) is nothing different with z_soc_irq_enable(irq) which still needs arm APIs.

```
void arch_irq_enable(unsigned int irq)
{
#ifdef CONFIG_MULTI_LEVEL_INTERRUPTS
        if (irq_get_level(irq) > 1)
                z_soc_irq_level2_enable(irq);
        else
#endif
        NVIC_EnableIRQ((IRQn_Type)irq);
}
```

So, I think current PR is okay to privide flexibility to soc.
Other intc drivers can still use irq_nextlevel.h, but that's not the case of irqsteer driver.
Thanks.

LaurentiuM1234 replied:
The multi-level interrupt support is already hard to deal with as it is. Introducing more architecture-specific APIs will only make things worse.

I believe that what this PR is doing is attempt to replicate what has been done for the XTENSA architecture for multi-level interrupts: provide functions for dealing with level 1 interrupts so that the aggregator's driver won't have to redefine them.

I understand that changing the architecture code may not be a good idea, but, at minimum, we still need to introduce a new condition in the architecture layer such that if CONFIG_MULTI_LEVEL_INTERRUPTS is enabled, we use the `z_soc_irq_*` suite (like it's done in the case of `CONFIG_ARM_CUSTOM_INTERRUPT_CONTROLLER`).

If the change proposed by @yangbolu1991 is not acceptable (i.e: have the level1 interrupt handling done by the z_arm_* suite) then maybe we can just do the level1 interrupt handling inside our driver?

e.g: for NVIC-based cores, define `irqsteer_level1_irq_enable` as `NVIC_EnableIRQ`. Same for GIC-based cores.

This is not ideal either since we'd need to introduce some additional checks but ATM I don't see another option?

Also, maybe I'm wrong here but I don't see much point to the `irq_nextlevel` API here? You'd still need to define the `z_soc_irq_*` functions?

wearyzen replied:
There is req created [here](https://github.com/zephyrproject-rtos/zephyr/issues/16168) to add multi level interrupt support with another driver but this hasn't been picked up yet so I was trying to see if we can have something that works as a generic solution.
I don't have any better suggestions atm so I'll be ok to changing the names just to be consistent with how this is done for Xtensa.

yangbolu1991 commented:
Sorry, a typo error on function name caused twister issues. Fixed.

dbaluta commented:
@yangbolu1991 IRQ Steer NXP part looks good to me.

wearyzen commented on include/zephyr/arch/arm/irq.h:
```diff
@@ -35,20 +35,26 @@ GTEXT(z_soc_irq_eoi)
+extern void arm_irq_disable(unsigned int irq);
+extern int arm_irq_is_enabled(unsigned int irq);
+extern void arm_irq_priority_set(unsigned int irq, unsigned int prio, uint32_t flags);
+#if !defined(CONFIG_MULTI_LEVEL_INTERRUPTS)
```
There is no need to have wrapper funcs for arch_irq, you can just redefine them as below:
```
#define arch_irq_enable(irq)		arm_irq_enable(irq)
#define arch_irq_disable(irq)		arm_irq_disable(irq)
#define arch_irq_is_enabled(irq)	arm_irq_is_enabled(irq)
```

yangbolu1991 commented:
Thank you very much @dbaluta  @wearyzen  and @LaurentiuM1234 
Updated to v4. Changes,
- Redefined arch_irq_xxx func for ARM interrupt control instead of using wrapper func.

Thanks.
